File: DSA/src/Questions/BST/LowestCommonAncestor.java
package Questions.BST;

import Questions.BinaryTree.TreeNode;

public class LowestCommonAncestor {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || p == root || q == root) {
            return root;
        }

        // Go to Left and Right
        TreeNode a = lowestCommonAncestor(root.left, p, q);
        TreeNode b = lowestCommonAncestor(root.right, p, q);

        if (a != null && b != null) {
            return root;
        }
        return (b == null) ? a : b;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        TreeNode left = new TreeNode(5);
        TreeNode right = new TreeNode(1);
        TreeNode leftLeft = new TreeNode(6);
        TreeNode leftRight = new TreeNode(2);
        TreeNode rightLeft = new TreeNode(0);
        TreeNode rightRight = new TreeNode(8);
        TreeNode leftRightLeft = new TreeNode(7);
        TreeNode leftRightRight = new TreeNode(4);
        root.left = left;
        root.right = right;
        left.left = leftLeft;
        left.right = leftRight;
        right.left = rightLeft;
        right.right = rightRight;
        leftRight.left = leftRightLeft;
        leftRight.right = leftRightRight;
        LowestCommonAncestor lowestCommonAncestor = new LowestCommonAncestor();
        TreeNode result = lowestCommonAncestor.lowestCommonAncestor(root, left, right);
        System.out.println("Lowest Common Ancestor: " + result.val);
    }
}

File: DSA/Python/Questions/BinaryTree/LowestCommonAncestor.py
"""
# Question: Lowest Common Ancestor of a Binary Tree
# Link: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/

# Time Complexity: O(N)
# Space Complexity: O(H)

# Algorithm:
# 1. DFS traversal
# 2. Find both nodes
# 3. Return LCA node
"""


class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None


class LowestCommonAncestor:
    def lowest_common_ancestor(
        self, root: TreeNode, p: TreeNode, q: TreeNode
    ) -> TreeNode:
        if not root or root == p or root == q:
            return root

        left = self.lowest_common_ancestor(root.left, p, q)
        right = self.lowest_common_ancestor(root.right, p, q)

        if left and right:
            return root
        return left if left else right


def main():
    solution = LowestCommonAncestor()
    root = TreeNode(3)
    root.left = TreeNode(5)
    root.right = TreeNode(1)
    print(solution.lowest_common_ancestor(root, root.left, root.right).val)


if __name__ == "__main__":
    main()
